### DISCUSSION TITLE
Accounting for non-uniformities in row geometry

### DISCUSSION THREAD

**Discussion by williamhobbs** [williamhobbs/2024_pvpmc_self_shade#4] [Ideas]
Even on a well-graded site, there will be some variation in PV row alignment. That will mean that calculations based on surface orientation and GCR, like shaded fraction (_fs_), might perfectly represent the mean, but not the variation across a site. 

I think this has to be a contributor to areas like the highlighted region in the chart below. Some rows have higher _fs_ than average, and some are lower. During transition periods where _fs_ is near zero, a little more shade is more bad than a little less shade is good.

![image](https://github.com/williamhobbs/2024_pvpmc_self_shade/assets/45701090/443bef23-59ff-4e44-a457-7efdaddbca13)

So, I think the net result is an underestimation of self-shade losses if the model is setup to exactly match site design geometry (starting to sound like subhourly clipping error, which can only hurt, not help...).

I can think of two possible solutions:
1. modify the modeled geometry to match something "worse" than the design geometry (e.g., increase the GCR a bit), or
2. have a model that takes the design geometry plus some variance/standard deviation value for one or more geometry parameters, then accounts for the issue more directly

For 2, one option I've thought about is having standard deviation of _fs_ as an input. Then calculate everything with sampled _fs_ values, e.g., 3 or 5 times, and combine results. You might need to calculate _fs_ in a way that allows for negative values, or otherwise figure out times when there is no possibility of shade.

**Comment by kandersolar** [comment]
I wonder how useful/accurate it would be to estimate the "actual" GCR and cross-axis slope by fitting the model to the measurements with `scipy.optimize.curve_fit` or similar.  

Also relevant to that highlighted difference is what we talked about the other day where "shaded module fraction" (what the geometry equations report) is not exactly the same as "shaded active area fraction", which is what matters for electrical modeling.  And another minor effect: the sun is not a point, which means its shadow is not a hard edge.  In fact when the math says $f_s$ is just about to go above zero, the bottom half of the sun is already occluded from the perspective of the bottom edge of the module.

**Reply by williamhobbs** [reply]
>estimate the "actual" GCR

I have found that to help some in previous work I've done.

>"shaded module fraction" (what the geometry equations report) is not exactly the same as "shaded active area fraction"

Yeah, good reminder. PVsyst has inputs for top+bottom (or left+right) "inactive bands" [1]. Not sure if it does anything for twin modules with half-cut cells like we mentioned. 

>And another minor effect: the sun is not a point

I thought about this earlier and forgot to include it in the post here. Seems like that could put a "floor" on the variance input for _fs_, maybe? 

[1] https://www.pvsyst.com/help/index.html?gcr.htm

**Reply by williamhobbs** [reply]
Here's a few GCR values (the first is the "real" value):

![image](https://github.com/williamhobbs/2024_pvpmc_self_shade/assets/45701090/ff975958-4861-4d29-8bb7-319116007dc7)

The real GCR seems to be the most consistent to me, matching on the "valleys" and overshooting the "peaks".
